picard/acousticbrainz/__init__.py

```diff
@@ -0,0 +1,121 @@
+# -*- coding: utf-8 -*-
+#
+# Picard, the next-generation MusicBrainz tagger
+#
+# Copyright 2014 Music Technology Group - Universitat Pompeu Fabra
+# Copyright 2020-2021 Gabriel Ferreira (@gabrielcarvfer)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 3
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+
+
+from functools import partial
+import json
+import os
+import subprocess
+import sys
+
+from picard.acousticbrainz.extractor import (
+    check_extractor_version,
+    precompute_extractor_sha,
+)
+from picard.util import load_json
+from picard.webservice import ratecontrol
+
+
+ACOUSTICBRAINZ_HOST = "acousticbrainz.org"
+ACOUSTICBRAINZ_PORT = 80
+
+ratecontrol.set_minimum_delay((ACOUSTICBRAINZ_HOST, ACOUSTICBRAINZ_PORT), 1000)
+
+acousticbrainz_extractor = os.path.dirname(sys.argv[0]) + os.sep + "streaming_extractor_music" + ("" if sys.platform != "win32" else ".exe")
+acousticbrainz_extractor_sha = precompute_extractor_sha(acousticbrainz_extractor)
+acousticbrainz_extractor_version = check_extractor_version(acousticbrainz_extractor)
+
+
+def run_extractor_entry(tagger, recording_id, input_path, output_path, extractor_callback):
+    tagger.webservice.get(ACOUSTICBRAINZ_HOST,
+                          ACOUSTICBRAINZ_PORT,
+                          "/%s/low-level" % recording_id,
+                          partial(run_extractor, tagger, input_path, output_path, extractor_callback),
+                          priority=True,
+                          important=False,
+                          parse_response_type=None)
+
+
+def extractor(input_path, output_path):
+    # If an a duplicate entry doesn't exist, process the file
+    args = [acousticbrainz_extractor, input_path, output_path]
+    startupinfo = subprocess.STARTUPINFO()
+    startupinfo.dwFlags |= subprocess.STARTF_USESHOWWINDOW

```

Comment:
Note that `subprocess.STARTUPINFO` is only available on Windows. What exactly does setting the `STARTF_USESHOWWINDOW` flag do, is that needed?

Reply:
It prevents new shells from popping up. I could use the same thing fpcalc does here.

Reply:
Ok, understood. I guess because fpcalc uses some Qt functions instead those take care of that already.

No objection to use subprocess here, especially as we have started to prefer Python standard library over Qt. I just would probably put it into a helper function in picard.util (maybe `run_executable` to follow the naming of `find_executable`) and the startupinfo stuff needs to be done only for Windows. Making it easy to call a command line executable and get back the output seems to be something that is very useful, also for some plugins.

Reply:
Both are fine for me. Changes are minimal.
https://github.com/metabrainz/picard/pull/1854/commits/052c04b62dfa14e7e2a7e5fd6d7e2384cb055691

Reply:
Ended up going the `run_executable` route.